cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/sequence.h>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+template <bool has_nulls, typename ReturnType = bool>
+struct unique_comparator {
+  unique_comparator(table_device_view device_table,
+                    size_type const *sorted_order)
+      : comp(device_table, device_table, true), perm(sorted_order) {}
+  __device__ ReturnType operator()(size_type index) const noexcept {
+    return index == 0 || not comp(perm[index], perm[index - 1]);
+  };
+
+private:
+  row_equality_comparator<has_nulls> comp;
+  size_type const *perm;
+};
+
+std::unique_ptr<table> rank(table_view const &input, rank_method method,
+                            order column_order, include_nulls _include_nulls,
+                            null_order null_precedence,
+                            bool percentage,
+                            rmm::mr::device_memory_resource *mr,
+                            cudaStream_t stream = 0) {
+  auto const size = input.num_rows();
+
+  std::vector<std::unique_ptr<column>> rank_columns;
+  for (auto const &input_col : input) {
```

Comment:
Wait a second, the documentation says that this function ranks rows lexicographically, but this looks like you're ranking each column independently! Those are very different things!

If we only care about ranking individual columns, then this function should just be a `column_view` API, not a table API. If it's a table API, then it should rank rows lexicographically.

Reply:
documentation should be updated. Each column is ordered independently. Making it as column only API make sense. I will update.

Reply:
changed to column API